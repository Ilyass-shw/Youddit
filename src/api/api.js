// A tiny wrapper around fetch(), borrowed from
// https://kentcdodds.com/blog/replace-axios-with-a-simple-custom-fetch-wrapper
import React from "react";
import fromUnixTime from "date-fns/fromUnixTime";
import { formatDistanceToNow } from "date-fns";
import millify from "millify";
import { Icon } from "@iconify/react";
import imageIcon from "@iconify/icons-bi/image";
import commentTextAlt from "@iconify/icons-zmdi/comment-text-alt";
import linkChain from "@iconify/icons-akar-icons/link-chain";

export async function client(endpoint, { body, ...customConfig } = {}) {
	// const headers = { 'Content-Type': 'application/json' }

	const config = {
		method: body ? "POST" : "GET",
		//   ...customConfig,
		//   headers: {
		//     ...headers,
		//     ...customConfig.headers,
		//   },
	};

	if (body) {
		config.body = JSON.stringify(body);
	}

	let data;
	try {
		const response = await window.fetch(endpoint, config);
		data = await response.json();
		if (response.ok) {
			return data;
		}
		throw new Error(response.statusText);
	} catch (err) {
		return Promise.reject(err.message ? err.message : data);
	}
}

client.get = function (endpoint, customConfig = {}) {
	return client(endpoint, { ...customConfig, method: "GET" });
};

client.post = function (endpoint, body, customConfig = {}) {
	return client(endpoint, { ...customConfig, body });
};

// ============= ============= ============= ============= =============

export const getEndPoint = (searchTerm, filter) => {
	return `https://www.reddit.com/search.json?q=${searchTerm}&limit=20&restrict_sr=true&sort=${filter}&t=all&show=all`;
};

// ============= ============= ============= ============= =============

export const handlefetchedPosts = (response) => {
	return response.data.children.map((post) => {
		let postDate = fromUnixTime(post.data.created_utc);

		postDate = formatDistanceToNow(postDate, { addSuffix: true, includeSeconds: true });

		const upvotes = millify(post.data.ups);

		let img;

		if (post.data.rpan_video) {
			img = post.data.rpan_video;
		} else if (post.data.thumbnail) {
			img = post.data.thumbnail;
		}

		return {
			title: post.data.title,
			img,
			url: post.data.permalink,
			upvotes,
			date: postDate,
			author: post.data.author,
			subreddit: post.data.subreddit_name_prefixed,
			id: post.data.id,
		};
	});
};

// ============= ============= ============= ============= =============

export const handleThumbnail = (thubmnailSrc) => {
	const style = {
		display: "inline-flex",
		justifyContent: "center",
		alignItems: "center",
		backgroundColor: "#202021",
	};

	const ImageIcon = (icon) => {
		return (
			<div className="content-image" style={style}>
				<Icon icon={icon} height="2rem" width="2rem" />
			</div>
		);
	};

	if (thubmnailSrc) {
		if (thubmnailSrc === "self") {
			return ImageIcon(commentTextAlt);
		} else if (thubmnailSrc === "spoiler") {
			return ImageIcon(imageIcon);
		} else if (thubmnailSrc === "default") {
			return ImageIcon(linkChain);
		} else if (thubmnailSrc === "image") {
			return ImageIcon(linkChain);
		} else {
			return <img src={thubmnailSrc} alt="post" className="content-image" />;
		}
	}
};
