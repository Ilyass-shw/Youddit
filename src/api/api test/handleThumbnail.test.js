import React from 'react'
import { handleThumbnail } from "../api";

describe('handleThumbnail',()=>{

    
    it('should return img element with te according provided link as src ', ()=>{
        const ThumbnailLink= 'www.yt.com'

        const expectedImgElement= <img src={ThumbnailLink} alt="post" className="content-image" />
        

        expect(handleThumbnail(ThumbnailLink)).toStrictEqual(expectedImgElement)
    })

    it('should return Icon if thumbnail=self ', ()=>{
        const ThumbnailLink= 'self'        

        expect(handleThumbnail(ThumbnailLink)).toMatchSnapshot()
    })

    it('should return Icon if thumbnail=spoiler ', ()=>{
        const ThumbnailLink= 'spoiler'        

        expect(handleThumbnail(ThumbnailLink)).toMatchSnapshot()
    })

    it('should return Icon if thumbnail=default ', ()=>{
        const ThumbnailLink= 'default'        

        expect(handleThumbnail(ThumbnailLink)).toMatchSnapshot()
    })

    it('should return Icon if thumbnail=image ', ()=>{
        const ThumbnailLink= 'image'        

        expect(handleThumbnail(ThumbnailLink)).toMatchSnapshot()
    })
})
