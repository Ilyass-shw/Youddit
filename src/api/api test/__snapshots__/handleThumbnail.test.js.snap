// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleThumbnail should return Icon if thumbnail=default  1`] = `
<div
  className="content-image"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#202021",
      "display": "inline-flex",
      "justifyContent": "center",
    }
  }
>
  <Icon
    height="2rem"
    icon={
      Object {
        "body": "<g fill=\\"none\\"><path d=\\"M13.544 10.456a4.368 4.368 0 0 0-6.176 0l-3.089 3.088a4.367 4.367 0 1 0 6.177 6.177L12 18.177m-1.544-4.633a4.368 4.368 0 0 0 6.177 0l3.088-3.088a4.368 4.368 0 0 0-6.177-6.177L12 5.823\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></g>",
        "height": 24,
        "width": 24,
      }
    }
    width="2rem"
  />
</div>
`;

exports[`handleThumbnail should return Icon if thumbnail=image  1`] = `
<div
  className="content-image"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#202021",
      "display": "inline-flex",
      "justifyContent": "center",
    }
  }
>
  <Icon
    height="2rem"
    icon={
      Object {
        "body": "<g fill=\\"none\\"><path d=\\"M13.544 10.456a4.368 4.368 0 0 0-6.176 0l-3.089 3.088a4.367 4.367 0 1 0 6.177 6.177L12 18.177m-1.544-4.633a4.368 4.368 0 0 0 6.177 0l3.088-3.088a4.368 4.368 0 0 0-6.177-6.177L12 5.823\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></g>",
        "height": 24,
        "width": 24,
      }
    }
    width="2rem"
  />
</div>
`;

exports[`handleThumbnail should return Icon if thumbnail=self  1`] = `
<div
  className="content-image"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#202021",
      "display": "inline-flex",
      "justifyContent": "center",
    }
  }
>
  <Icon
    height="2rem"
    icon={
      Object {
        "body": "<path d=\\"M384 3q18 0 30.5 12.5T427 45v256q0 18-12.5 30.5T384 344H85L0 429V45q0-17 12.5-29.5T43 3h341zM85 152v43h256v-43H85zm171 107v-43H85v43h171zm85-128V88H85v43h256z\\" fill=\\"currentColor\\"/>",
        "height": 432,
        "width": 432,
      }
    }
    width="2rem"
  />
</div>
`;

exports[`handleThumbnail should return Icon if thumbnail=spoiler  1`] = `
<div
  className="content-image"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#202021",
      "display": "inline-flex",
      "justifyContent": "center",
    }
  }
>
  <Icon
    height="2rem"
    icon={
      Object {
        "body": "<g fill=\\"currentColor\\"><path d=\\"M6.002 5.5a1.5 1.5 0 1 1-3 0a1.5 1.5 0 0 1 3 0z\\"/><path d=\\"M2.002 1a2 2 0 0 0-2 2v10a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V3a2 2 0 0 0-2-2h-12zm12 1a1 1 0 0 1 1 1v6.5l-3.777-1.947a.5.5 0 0 0-.577.093l-3.71 3.71l-2.66-1.772a.5.5 0 0 0-.63.062L1.002 12V3a1 1 0 0 1 1-1h12z\\"/></g>",
        "height": 16,
        "width": 16,
      }
    }
    width="2rem"
  />
</div>
`;
